ring/partition_ring_model.go

```diff
@@ -0,0 +1,336 @@
+	m.Owners[id] = updated
+	return true
+}
+
+// RemoveOwner removes a partition owner.
+func (m *PartitionRingDesc) RemoveOwner(id string) {
+	delete(m.Owners, id)
+}
+
+// HasOwner returns whether a owner exists.
+func (m *PartitionRingDesc) HasOwner(id string) bool {
+	_, ok := m.Owners[id]
+	return ok
+}
+
+// Merge implements memberlist.Mergeable.
+func (m *PartitionRingDesc) Merge(mergeable memberlist.Mergeable, localCAS bool) (memberlist.Mergeable, error) {
+	return m.mergeWithTime(mergeable, localCAS, time.Now())
+}
+
+func (m *PartitionRingDesc) mergeWithTime(mergeable memberlist.Mergeable, localCAS bool, now time.Time) (memberlist.Mergeable, error) {
+	if mergeable == nil {
+		return nil, nil
+	}
+
+	other, ok := mergeable.(*PartitionRingDesc)
+	if !ok {
+		// This method only deals with non-nil rings.
+		return nil, fmt.Errorf("expected *PartitionRingDesc, got %T", mergeable)
+	}
+
+	if other == nil {
+		return nil, nil
+	}
+
+	change := NewPartitionRingDesc()
+
+	// Handle partitions.
+	for id, otherPart := range other.Partitions {
+		changed := false
+
+		thisPart, exists := m.Partitions[id]
+		if !exists {
+			changed = true
+			thisPart = otherPart
+		} else {
+			// We expect tokens to be immutable, so we don't need to check for conflicts.
+			//
+			// If in the future we'll change the tokens generation algorithm and we'll have to handle migration to
+			// a different set of tokens then we'll add the support. For example, we could add "token generation version"
+			// to PartitionDesc and then preserve tokens generated by latest version only.
+			if len(thisPart.Tokens) == 0 && len(otherPart.Tokens) > 0 {
+				thisPart.Tokens = otherPart.Tokens
+				changed = true
+			}
+
+			if otherPart.StateTimestamp > thisPart.StateTimestamp {
+				changed = true
+
+				thisPart.State = otherPart.State
+				thisPart.StateTimestamp = otherPart.StateTimestamp
+			}
+		}
+
+		if changed {
+			m.Partitions[id] = thisPart
+			change.Partitions[id] = thisPart
+		}
+	}
+
+	if localCAS {
+		// Let's mark all missing partitions in incoming change as deleted.
+		// This breaks commutativity! But we only do it locally, not when gossiping with others.
+		for pid, thisPart := range m.Partitions {
+			if _, exists := other.Partitions[pid]; !exists && thisPart.State != PartitionDeleted {
+				// Partition was removed from the ring. We need to preserve it locally, but we set state to PartitionDeleted.
+				thisPart.State = PartitionDeleted
+				thisPart.StateTimestamp = now.Unix()
+				m.Partitions[pid] = thisPart
+				change.Partitions[pid] = thisPart
+			}
+		}
+	}
+
+	// Now let's handle owners. Owners don't have tokens, which simplifies things compared to "normal" ring.
+	for id, otherOwner := range other.Owners {
+		thisOwner := m.Owners[id]
+
+		if otherOwner.UpdatedTimestamp > thisOwner.UpdatedTimestamp || (otherOwner.UpdatedTimestamp == thisOwner.UpdatedTimestamp && otherOwner.State == OwnerDeleted && thisOwner.State != OwnerDeleted) {
```

Comment:
For owners we accept "Deleted" state even if timestamp hasn't changed, but for partitions we don't. 

Two comments:
* I'd suggest to be consistent for both partitions and owners -- either we keep the check or remove it. To be honest, I don't remember why this case exists in the original Ring code. I think it would be fine to remove this special check.
* If we decide to keep this check, let's add test for it.

Reply:
> To be honest, I don't remember why this case exists in the original Ring code.

Maybe to cover the case it was removed in less than 1s from the last update? The timestamp has second precision.

Reply:
I think the reason is that original ring code did not update timestamp when setting entry state to LEFT (tombstone). See `Merge` method in https://github.com/cortexproject/cortex/pull/1727/files#diff-e773119be1c2446745f5b5eeb2d7b436e99cb21c02e1fdfc40e42d041d541ad8 and how `localCAS` is handled.

Your point about 1s precision is good reason why we may want to keep it.